membership-attribute-service/app/monitoring/Metrics.scala

```diff
@@ -1,9 +1,40 @@
 package monitoring
 
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+import akka.actor.ActorSystem
 import com.gu.monitoring.CloudWatch
+import com.typesafe.scalalogging.LazyLogging
 import configuration.Config
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
 
 case class Metrics(service: String) extends CloudWatch {
   val stage = Config.stage
   val application = Config.applicationName // This sets the namespace for Custom Metrics in AWS (see CloudWatch)
 }
+
+// Designed for high-frequency metrics, for example, 1000 per minute is about $400 per month
+final class ExpensiveMetrics(
+  val service: String
+)(implicit system: ActorSystem, ec: ExecutionContext) extends CloudWatch with LazyLogging {
+  import scala.collection.JavaConverters._
+  private val chm = new ConcurrentHashMap[String, AtomicInteger]().asScala // keep it first in the constructor
+
+  val stage = Config.stage
+  val application = Config.applicationName
+
+  system.scheduler.schedule(5.seconds, 60.seconds)(publishAllMetrics())
+
+  def countRequest(key: String): Unit =
+    chm.getOrElseUpdate(key, new AtomicInteger(1)).incrementAndGet()
```

Comment:
https://42ways.de/blog/2015/02/21/scala-pitfalls-concurrent-hashmap/

Reply:
Not sure this is relevant anymore since Scala 2.12: https://github.com/scala/scala/pull/4323 I think implementation of [concurrent.Map#getOrElseUpdate](https://github.com/scala/scala/blob/cd8410d30427747bbcd1c94429ce1509e527a315/src/library/scala/collection/concurrent/Map.scala#L94) uses `putIfAbsent` which documentation states is [atomic](https://docs.scala-lang.org/overviews/collections/concrete-mutable-collection-classes.html#concurrent-maps).